pageserver/src/task_mgr.rs

```diff
@@ -481,13 +481,27 @@ pub async fn shutdown_tasks(
     for task in victim_tasks {
         let join_handle = {
             let mut task_mut = task.mutable.lock().unwrap();
-            info!("waiting for {} to shut down", task.name);
-            let join_handle = task_mut.join_handle.take();
-            drop(task_mut);
-            join_handle
+            task_mut.join_handle.take()
         };
-        if let Some(join_handle) = join_handle {
-            let _ = join_handle.await;
+        if let Some(mut join_handle) = join_handle {
+            // let's see if test failures could be caused by allocating the timeout in most cases
+            if futures::future::poll_immediate(&mut join_handle)
+                .await
+                .is_some()
+            {
+                return;
+            }
+
+            let mut passed = false;
+            loop {
+                tokio::select! {
+                    _ = &mut join_handle => { return; },
```

Comment:
These should be continues, not returns :)

Reply:
[bf1d055](https://github.com/neondatabase/neon/pull/3866/commits/bf1d055c0e85113a22b8c426f5864f82805a86ee)